aten/src/ATen/native/mkl/SparseCsrLinearAlgebra.cpp

```diff
@@ -0,0 +1,217 @@
+#if !AT_MKL_ENABLED() || defined(_MSC_VER) || defined(__APPLE__) || defined(__MACH__)
+
+namespace at {
+namespace native {
+using namespace at::sparse;
+Tensor& _sparse_mm_mkl_(
+    Tensor& self,
+    const SparseTensor& sparse_,
+    const Tensor& dense,
+    const Tensor& t,
+    const Scalar& alpha,
+    const Scalar& beta) {
+#if _MSC_VER
+  AT_ERROR("sparse_mm_mkl: MKL support is disabled on Windows");
+#elif __APPLE__ || __MACH__
+  AT_ERROR("sparse_mm_mkl: MKL support is disabled on macos/iOS.");
+#else
+  AT_ERROR("sparse_mm_mkl: ATen not compiled with MKL support");
+#endif
+}
+} // namespace native
+} // namespace at
+
+#else // AT_MKL_ENABLED
+
+#include <ATen/mkl/Descriptors.h>
+#include <ATen/mkl/Exceptions.h>
+#include <ATen/mkl/Limits.h>
+#include <mkl.h>
+#include <mkl_spblas.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/SparseCsrTensorImpl.h>
+
+#ifdef MKL_ILP64
+#define TORCH_INT_TYPE at::kLong
+#else
+#define TORCH_INT_TYPE at::kInt
+#endif
+
+namespace at {
+namespace native {
+using namespace at::sparse;
+
+static inline void sparse_mm_mkl_impl(
+    float* res,
+    MKL_INT* col_indices,
+    MKL_INT* crow_indices,
+    float* values,
+    float* dense,
+    float* t,
+    float alpha,
+    float beta,
+    MKL_INT nrows,
+    MKL_INT ncols,
+    MKL_INT dense_ncols) {
+  sparse_matrix_t A = 0;
+  matrix_descr desc;
+  desc.type = SPARSE_MATRIX_TYPE_GENERAL;
+  int retval = mkl_sparse_s_create_csr(
+      &A,
+      SPARSE_INDEX_BASE_ZERO,
+      nrows,
+      ncols,
+      crow_indices,
+      crow_indices + 1,
+      col_indices,
+      values);
+  TORCH_CHECK(
+      retval == 0, "mkl_sparse_s_create_csr failed with error code: ", retval);
+
+  mkl_sparse_s_mm(
+      SPARSE_OPERATION_NON_TRANSPOSE,
+      alpha,
+      A,
+      desc,
+      SPARSE_LAYOUT_ROW_MAJOR,
+      dense,
+      dense_ncols,
+      dense_ncols,
+      beta,
+      res,
+      dense_ncols);
+  mkl_sparse_destroy(A);
+}
+
+static inline void sparse_mm_mkl_impl(
+    double* res,
+    MKL_INT* col_indices,
+    MKL_INT* crow_indices,
+    double* values,
+    double* dense,
+    double* t,
+    double alpha,
+    double beta,
+    MKL_INT nrows,
+    MKL_INT ncols,
+    MKL_INT dense_ncols) {
+  sparse_matrix_t A = 0;
+  matrix_descr desc;
+  desc.type = SPARSE_MATRIX_TYPE_GENERAL;
+  int retval = mkl_sparse_d_create_csr(
+      &A,
+      SPARSE_INDEX_BASE_ZERO,
+      nrows,
+      ncols,
+      crow_indices,
+      crow_indices + 1,
+      col_indices,
+      values);
+  TORCH_CHECK(
+      retval == 0, "mkl_sparse_d_create_csr failed with error code: ", retval);
+
+  mkl_sparse_d_mm(
+      SPARSE_OPERATION_NON_TRANSPOSE,
+      alpha,
+      A,
+      desc,
+      SPARSE_LAYOUT_ROW_MAJOR,
+      dense,
+      dense_ncols,
+      dense_ncols,
+      beta,
+      res,
+      dense_ncols);
+  mkl_sparse_destroy(A);
```

Comment:
I'd probably suggest making a small C++ RAII class to handle deallocation. As written today this function cannot leak A but it is very easy to accidentally leak

Reply:
Fixed in master.